ipld/plugin/nmt.go

```diff
@@ -53,9 +53,24 @@ func NewNamespaceHasher(hasher *nmt.Hasher) hash.Hash {
 }
 
 func (n *namespaceHasher) Write(data []byte) (int, error) {
-	n.tp = data[0]
+	ln, nln, hln := len(data), int(n.NamespaceLen), n.Hash.Size()
+	switch ln {
+	default:
+		return 0, fmt.Errorf("wrong data size")
+	// inner node
+	case (nln*2+hln)*2 + 1: // with type byte
+		fallthrough
+	case (nln*2 + hln) * 2:
+		n.tp = nmt.NodePrefix
+	// leaf data node
+	case hln + 256:
+		fallthrough
+	case hln + 256 + 1: // with type byte
+		n.tp = nmt.LeafPrefix
+	}
```

Comment:
Note that there 4 cases, as I made this code in preparation for type byte removal. Having 4 cases here would allow us to remove the type byte + ensure that existing data on the network continues to work